codex/utils/statemachine.nim

```diff
@@ -100,3 +107,24 @@ method enter(state: AsyncState) =
 
 method exit(state: AsyncState) =
   asyncSpawn state.exitAsync()
+
+proc switchAsync*(machine: StateMachineAsync, newState: AsyncState) {.async.} =
+  if state =? (machine.state as AsyncState):
+    trace "Switching sales state", `from`=state, to=newState
+    if activeTransition =? state.activeTransition and
+        not activeTransition.completed:
+      await activeTransition.cancelAndWait()
+    # should wait for exit before switch. could add a transition option during
+    # switch if we don't need to wait
+    await state.exitAsync()
+    state.context = none StateMachine
+  else:
+    trace "Switching sales state", `from`="no state", to=newState
```

Comment:
This seems overly specific for the generic state machine

Reply:
Removed "sales ".